agent/eventstream/eventstream.go

```diff
@@ -0,0 +1,106 @@
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package handler deals with appropriately reacting to all ACS messages as well
+// as maintaining the connection to ACS.
+package eventstream
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/cihub/seelog"
+	"golang.org/x/net/context"
+)
+
+type eventHandler func(...interface{}) error
+
+// EventStream waiting for events and notifying the listeners by invoking
+// the handler that listeners registered
+type EventStream struct {
+	name         string
+	event        chan interface{}
+	stopped      chan struct{}
+	handlers     []eventHandler
+	ctx          context.Context
+	handlersLock sync.RWMutex
+}
+
+func NewEventStream(name string, ctx context.Context) *EventStream {
+	return &EventStream{
+		name:    name,
+		event:   make(chan interface{}, 1),
+		stopped: make(chan struct{}, 1),
+		ctx:     ctx,
+	}
+}
+
+// Subscribe adds the handler to be called into EventStream
+func (eventStream *EventStream) Subscribe(handler eventHandler) {
+	eventStream.handlersLock.Lock()
+	defer eventStream.handlersLock.Unlock()
+
+	eventStream.handlers = append(eventStream.handlers, handler)
+}
+
+// broadcast calls all handler's handler function
+func (eventStream *EventStream) broadcast(event interface{}) {
+	eventStream.handlersLock.RLock()
+	defer eventStream.handlersLock.RUnlock()
+
+	seelog.Debugf("Event stream %v received events, broadcasting to listeners...", eventStream.name)
+
+	for _, handlerFunc := range eventStream.handlers {
+		go handlerFunc(event)
+	}
+}
+
+// Unsubscribe deletes the handler from the EventStream
+func (eventStream *EventStream) Unsubscribe(handlerToUnsubscribe eventHandler) {
+	eventStream.handlersLock.Lock()
+	defer eventStream.handlersLock.Unlock()
+
+	for i, handler := range eventStream.handlers {
+		if reflect.ValueOf(handler).Pointer() == reflect.ValueOf(handlerToUnsubscribe).Pointer() {
+			eventStream.handlers = append(eventStream.handlers[:i], eventStream.handlers[i+1:]...)
+			return
+		}
+	}
+}
+
+// EventChannel returns the channel the event stream is listening
+func (eventStream *EventStream) EventChannel() chan interface{} {
+	return eventStream.event
+}
+
+// Context returns the context of event stream
+func (eventStream *EventStream) Context() context.Context {
+	return eventStream.ctx
+}
+
+// StartListening start listening to the event stream and handle received event
+func (eventStream *EventStream) StartListening() {
+	go func() {
+		seelog.Infof("Event stream %v start listening...", eventStream.name)
+		for {
+			select {
+			case event := <-eventStream.event:
+				eventStream.broadcast(event)
+			case <-eventStream.ctx.Done():
+				seelog.Infof("Event stream %v stopped listening...", eventStream.name)
+				close(eventStream.stopped)
```

Comment:
This is risky.  If somehow the context is canceled but something still tries to write into the channel, this will cause a panic.  It might be better to have the channel completely private in the event stream and expose a method to write into it that can check whether the stream is open or not and return an error when closed instead of a panic.